### PR TITLE
ssl tests

### PR DESCRIPTION
**Pull request by joedevivo**
This PR lets us get some https testing of vegur done. There were some edge cases that were pattern matching on `ranch_proxy` that also needed to match on `ranch_proxy_ssl`. That was fixed in a different PR (#151), but this will give us coverage going forward.

### CODE REVIEW

File: test/vegur_proxy_SUITE.erl [hunk 1]
```diff
@@ -68,19 +72,28 @@ init_per_suite(Config) ->
...
 
+init_per_group(http, Config) ->
+    VegurPort = 9333,
+    {ok, _} = vegur:start_http(VegurPort, vegur_stub, []),
+    mock_through(),
+    [{vegur_port, VegurPort}, {scheme, "http"} | Config];
+init_per_group(https, Config) ->
+    VegurPort = 9333,
+    SSLOptions = [{certfile,   "../../../../test/localhost.crt"},
+                  {keyfile,    "../../../../test/localhost.key"},
```
Comment (ypaq) [inline test/vegur_proxy_SUITE.erl]: Can you get these files via *_SUITE_data?

Comment (joedevivo) [inline test/vegur_proxy_SUITE.erl]: I could. I'm always a bit conflicted on this, because I wind up using them in more than one suite

**Comment by joedevivo** [comment]
This was working fine by itself, but when run with just `rebar3 ct`, some things were skipping. will dig in to that tomorrow.

File: test/vegur_proxy_SUITE.erl (hunk 1 above)
Comment (ferd) [inline test/vegur_proxy_SUITE.erl]: I'd just copy the _data dir for many suites in this case. I'd rather have the data dir working since that makes for more portable test runs

File: src/vegur.erl [hunk 2]
```diff
@@ -78,6 +81,17 @@ start_http(Port, Interface, Config) ->
...
     start(http, HttpRef, Port, Interface, Config).
 
+-spec start_https(PortNumber, Interface, Options, SSLOptions) ->
+                        {ok, pid()}|no_return() when
+      PortNumber :: inet:port_number(),
+      Interface :: module(),
+      Options :: options(),
+      SSLOptions :: [ssl:ssloption()].
+start_https(Port, Interface, Config, SSLOptions) ->
+    ok = set_config(extra_socket_options, SSLOptions ++ extra_socket_options()),
```
Comment (ferd) [inline src/vegur.erl]: Why is this required? 
I think there is a risk that this pollutes SSL socket options and uses them for non-SSL listeners depending on the start order.

File: src/vegur.erl [hunk 3]
```diff
@@ -138,31 +156,58 @@ defaults() ->
...
+       {route_time_header, route_time_header,
+        vegur_utils:config(route_time_header)},
+       {request_id_max_size, request_id_max_size,
+        vegur_utils:config(request_id_max_size)},
+       {downstream_connect_timeout,
+        downstream_connect_timeout,
+        vegur_utils:config(downstream_connect_timeout)},
+       acceptors,
+       max_connections
+      ]).
```
Comment (ferd) [inline src/vegur.erl]: I don't understand the intent behind that part of the diff. It looks like it's called only once and now takes twice the number of lines + a fold to do the same thing it did?

File: test/vegur_proxy_SUITE.erl [hunk 4]
```diff
@@ -90,6 +103,12 @@ init_per_group(vegur_proxy_connect, Config) ->
     mock_backend(9555),
     [{backend_port, BackendPort} | Config].
 
+end_per_group(http, Config) ->
+    vegur:stop_http(),
+    Config;
+end_per_group(https, Config) ->
+    vegur:stop_http(),
```
Comment (ferd) [inline test/vegur_proxy_SUITE.erl]: shouldn't that be `stop_https` ?

File: src/vegur.erl (hunk 2 above)
Comment (joedevivo) [inline src/vegur.erl]: oh I see, because extra_socket_options is an application environment thing?

Comment (ferd) [inline src/vegur.erl]: Yeah, and it's shared by non-https handlers too.

File: src/vegur.erl (hunk 3 above)
Comment (joedevivo) [inline src/vegur.erl]: I saw us doing something similar inside snit. I thought it was just a cleaner way that we preferred here.

File: test/vegur_proxy_SUITE.erl (hunk 4 above)
Comment (joedevivo) [inline test/vegur_proxy_SUITE.erl]: yup

File: src/vegur.erl (hunk 3 above)
Comment (ferd) [inline src/vegur.erl]: snit uses it because some values are to override default ones -- impossible to change them. I don't think this is required to the same extent here where we allow the user to override everything?

File: test/vegur_proxy_SUITE.erl (hunk 4 above)
Comment (joedevivo) [inline test/vegur_proxy_SUITE.erl]: That was also what was breaking my tests in the `rebar3 ct` run. So, thanks! :D

**Comment by ferd** [comment]
+1 as long as the options still work! (I don't see why they wouldn'T)
